api/src/main/java/org/motechproject/nms/api/web/KilkariController.java

```diff
@@ -101,4 +104,73 @@ public void saveInboxCallDetails(@RequestBody InboxCallDetailsRequest request) {
         //todo: the work...
     }
 
+    /**
+     * 4.2.3
+     * Create Subscription
+     *
+     * IVR shall invoke this API to create a new Kilkari subscription
+     *
+     */
+    @RequestMapping(value = "/subscription",
+            method = RequestMethod.POST,
+            headers = { "Content-type=application/json" })
+    @ResponseStatus(HttpStatus.OK)
+    public void createSubscription(@RequestBody SubscriptionRequest subscriptionRequest)
+        throws NotFoundException {
+
+        StringBuilder failureReasons = validate(subscriptionRequest.getCallingNumber(),
+                subscriptionRequest.getOperator(), subscriptionRequest.getCircle(),
+                subscriptionRequest.getCallId());
+        validateFieldPresent(failureReasons, "subscriptionPack", subscriptionRequest.getSubscriptionPack());
+        validateFieldPresent(failureReasons, "languageLocationCode",
+                subscriptionRequest.getLanguageLocationCode().toString());
+
+        if (failureReasons.length() > 0) {
+            throw new IllegalArgumentException(failureReasons.toString());
+        }
+        if (!validateSubscriptionPack(subscriptionRequest.getSubscriptionPack())) {
+            throw new NotFoundException(String.format(NOT_FOUND, "subscriptionPack"));
+        }
+
+        kilkariService.createSubscription(subscriptionRequest.getCallingNumber(),
+                subscriptionRequest.getLanguageLocationCode(), subscriptionRequest.getSubscriptionPack());
+    }
+
+    /**
+     * 4.2.4
+     * Deactivate Subscription
+     *
+     * IVR shall invoke this API to deactivate an existing Kilkari subscription
+     *
+     */
+    @RequestMapping(value = "/subscription",
+            method = RequestMethod.DELETE,
+            headers = { "Content-type=application/json" })
+    @ResponseStatus(HttpStatus.OK)
+    public void deactivateSubscription(@RequestBody SubscriptionRequest subscriptionRequest)
+        throws NotFoundException {
+
+        StringBuilder failureReasons = validate(subscriptionRequest.getCallingNumber(),
+                subscriptionRequest.getOperator(), subscriptionRequest.getCircle(),
+                subscriptionRequest.getCallId());
+
+        validateFieldPresent(failureReasons, "subscriptionId", subscriptionRequest.getSubscriptionId());
+
+        if (failureReasons.length() > 0) {
+            throw new IllegalArgumentException(failureReasons.toString());
+        }
+
+        Subscription subscription = kilkariService.getSubscription(subscriptionRequest.getSubscriptionId());
+
+        if (subscription == null) {
+            throw new NotFoundException(String.format(NOT_FOUND, "subscriptionId"));
+        }
+
+        kilkariService.deactivateSubscription(subscription);
+    }
+
+    private boolean validateSubscriptionPack(String name) {
+        return kilkariService.getSubscriptionPack(name) != null;
```

Comment:
File this under premature optimization but if you are loading the record just to validate the name, you can use the _count lookup so less data is sent and the object doesn't get constructed.